# The Verdict book is taken as the sample dataset here.

#### Step 1 : Creating tokens

In [88]:
with open("the-verdict.txt","r",encoding="utf-8") as f:
    raw_text = f.read()
print("Total number of characters : ",len(raw_text))
print(raw_text[:99])

Total number of characters :  20479
I HAD always thought Jack Gisburn rather a cheap genius--though a good fellow enough--so it was no 


###### Tokenising with whitespaces

In [39]:
import re

text="Hello World. Sample Testing Sentence."
result = re.split(r'(\s)',text)
print(result)

['Hello', ' ', 'World.', ' ', 'Sample', ' ', 'Testing', ' ', 'Sentence.']


###### Tokenising with whitespaces and punctuation marks - (,.)

In [40]:
result = re.split(r'([,.]|\s)',text)
print(result)

['Hello', ' ', 'World', '.', '', ' ', 'Sample', ' ', 'Testing', ' ', 'Sentence', '.', '']


###### Removing whitespaces

In [41]:
result=[item for item in result if item.strip()]
print(result)

['Hello', 'World', '.', 'Sample', 'Testing', 'Sentence', '.']


###### Removing punctuation marks like ,.:;?_!"()\' and whitespaces

In [42]:
text="Hello, World. Is this-- a test?"
result=re.split(r'([,.:;?_!"()\']|--|\s)',text)
result=[item for item in result if item.strip()]
print(result)

['Hello', ',', 'World', '.', 'Is', 'this', '--', 'a', 'test', '?']


###### Applying tokenization on the sample text "The Verdict"

In [43]:
preprocessed=re.split(r'([,.:;?_!"()\']|--|\s)',raw_text)
preprocessed=[item for item in preprocessed if item.strip()]
print(preprocessed)

['I', 'HAD', 'always', 'thought', 'Jack', 'Gisburn', 'rather', 'a', 'cheap', 'genius', '--', 'though', 'a', 'good', 'fellow', 'enough', '--', 'so', 'it', 'was', 'no', 'great', 'surprise', 'to', 'me', 'to', 'hear', 'that', ',', 'in', 'the', 'height', 'of', 'his', 'glory', ',', 'he', 'had', 'dropped', 'his', 'painting', ',', 'married', 'a', 'rich', 'widow', ',', 'and', 'established', 'himself', 'in', 'a', 'villa', 'on', 'the', 'Riviera', '.', '(', 'Though', 'I', 'rather', 'thought', 'it', 'would', 'have', 'been', 'Rome', 'or', 'Florence', '.', ')', '"', 'The', 'height', 'of', 'his', 'glory', '"', '--', 'that', 'was', 'what', 'the', 'women', 'called', 'it', '.', 'I', 'can', 'hear', 'Mrs', '.', 'Gideon', 'Thwing', '--', 'his', 'last', 'Chicago', 'sitter', '--', 'deploring', 'his', 'unaccountable', 'abdication', '.', '"', 'Of', 'course', 'it', "'", 's', 'going', 'to', 'send', 'the', 'value', 'of', 'my', 'picture', "'", 'way', 'up', ';', 'but', 'I', 'don', "'", 't', 'think', 'of', 'that', ',

In [44]:
print("Number of tokens in the sample : ",len(preprocessed))

Number of tokens in the sample :  4690


#### Step 2 : Creating tokenIDs

In [45]:
# Sorting the tokens alphabetically and taking only unique tokens
all_words = sorted(set(preprocessed))
vocab_size=len(all_words)
print("Length of unique tokens : ",vocab_size)

Length of unique tokens :  1130


In [46]:
vocab = {token:integer for integer,token in enumerate(all_words)}
print(vocab)

{'!': 0, '"': 1, "'": 2, '(': 3, ')': 4, ',': 5, '--': 6, '.': 7, ':': 8, ';': 9, '?': 10, 'A': 11, 'Ah': 12, 'Among': 13, 'And': 14, 'Are': 15, 'Arrt': 16, 'As': 17, 'At': 18, 'Be': 19, 'Begin': 20, 'Burlington': 21, 'But': 22, 'By': 23, 'Carlo': 24, 'Chicago': 25, 'Claude': 26, 'Come': 27, 'Croft': 28, 'Destroyed': 29, 'Devonshire': 30, 'Don': 31, 'Dubarry': 32, 'Emperors': 33, 'Florence': 34, 'For': 35, 'Gallery': 36, 'Gideon': 37, 'Gisburn': 38, 'Gisburns': 39, 'Grafton': 40, 'Greek': 41, 'Grindle': 42, 'Grindles': 43, 'HAD': 44, 'Had': 45, 'Hang': 46, 'Has': 47, 'He': 48, 'Her': 49, 'Hermia': 50, 'His': 51, 'How': 52, 'I': 53, 'If': 54, 'In': 55, 'It': 56, 'Jack': 57, 'Jove': 58, 'Just': 59, 'Lord': 60, 'Made': 61, 'Miss': 62, 'Money': 63, 'Monte': 64, 'Moon-dancers': 65, 'Mr': 66, 'Mrs': 67, 'My': 68, 'Never': 69, 'No': 70, 'Now': 71, 'Nutley': 72, 'Of': 73, 'Oh': 74, 'On': 75, 'Once': 76, 'Only': 77, 'Or': 78, 'Perhaps': 79, 'Poor': 80, 'Professional': 81, 'Renaissance': 82, 'Ri

In [47]:
for i,item in enumerate(vocab.items()):
    print(item)
    if i>=50:
        break

('!', 0)
('"', 1)
("'", 2)
('(', 3)
(')', 4)
(',', 5)
('--', 6)
('.', 7)
(':', 8)
(';', 9)
('?', 10)
('A', 11)
('Ah', 12)
('Among', 13)
('And', 14)
('Are', 15)
('Arrt', 16)
('As', 17)
('At', 18)
('Be', 19)
('Begin', 20)
('Burlington', 21)
('But', 22)
('By', 23)
('Carlo', 24)
('Chicago', 25)
('Claude', 26)
('Come', 27)
('Croft', 28)
('Destroyed', 29)
('Devonshire', 30)
('Don', 31)
('Dubarry', 32)
('Emperors', 33)
('Florence', 34)
('For', 35)
('Gallery', 36)
('Gideon', 37)
('Gisburn', 38)
('Gisburns', 39)
('Grafton', 40)
('Greek', 41)
('Grindle', 42)
('Grindles', 43)
('HAD', 44)
('Had', 45)
('Hang', 46)
('Has', 47)
('He', 48)
('Her', 49)
('Hermia', 50)


###### Making a tokenizer class for encoding and decoding

In [64]:
class SimpleTokenizerV1:
    def __init__(self,vocab):
        self.str_to_int=vocab
        self.int_to_str = {integer:token for token,integer in vocab.items()}
        
    #Encoder returns the tokenIDs for the input text    
    def encode(self,text):
        preprocessed = re.split(r'([,.:;?_!"()\']|--|\s)',text)
        preprocessed = [item for item in preprocessed if item.strip()]
        ids = [self.str_to_int[token] for token in preprocessed]
        return ids
    
    #Decoder returns the output sentence where the input is the tokenIDs
    def decode(self,ids):
        output_string=" ".join([self.int_to_str[i] for i in ids])
        # Replace spaces before the specified punctuations
        output_string=re.sub(r'\s+(--|[,.?!"();\'])',r'\1',output_string)
        return output_string
        

###### Instantiate a new tokenizer object from SimpleTokenizerV1 class and tokenize a passage from "The Verdict"

In [65]:
sample_text = """I have mentioned-- that Mrs. Gisburn was rich; 
                 and it was immediately perceptible that her husband was 
                 extracting from this circumstance a delicate but substantial 
                 satisfaction. It is, as a rule, the people 
                 who scorn money who get most out of it; and 
                 Jack's elegant disdain of his wife's big balance enabled him, 
                 with an appearance of perfect good-breeding, to transmute it into objects of art and luxury."""

tokenizer = SimpleTokenizerV1(vocab)

#Encoding the sample passage from the training dataset
ids=tokenizer.encode(sample_text)
print(ids)

#Decoding the ids to get the initial input sentence
print(tokenizer.decode(ids))

[53, 530, 667, 6, 987, 67, 7, 38, 1077, 841, 9, 157, 585, 1077, 567, 764, 987, 539, 562, 1077, 414, 477, 999, 266, 115, 323, 239, 947, 853, 7, 56, 584, 5, 177, 115, 848, 5, 988, 763, 1097, 859, 681, 1097, 488, 686, 738, 722, 585, 9, 157, 57, 2, 850, 384, 346, 722, 549, 1103, 2, 850, 219, 193, 387, 546, 5, 1108, 156, 166, 722, 765, 501, 5, 1016, 1027, 585, 579, 719, 722, 174, 157, 652, 7]
I have mentioned-- that Mrs. Gisburn was rich; and it was immediately perceptible that her husband was extracting from this circumstance a delicate but substantial satisfaction. It is, as a rule, the people who scorn money who get most out of it; and Jack' s elegant disdain of his wife' s big balance enabled him, with an appearance of perfect good-breeding, to transmute it into objects of art and luxury.


##### The above tokenizer will not work for the tokens out of the vocabulary.

In [66]:
sample_text="hello, world!"
ids=tokenizer.encode(sample_text)

KeyError: 'hello'

##### Adding <|endoftext\> and <|unk|> tokens to vocabulary.

In [69]:
max_index=max(vocab.values())
vocab["<|endoftext|>"]=max_index+1
vocab["<|unk|>"]=max_index+2

In [70]:
len(vocab.items())

1132

In [71]:
for i,item in enumerate(list(vocab.items())[-5:]):
    print(item)

('younger', 1127)
('your', 1128)
('yourself', 1129)
('<|endoftext|>', 1131)
('<|unk|>', 1132)


##### Tokenizer Class V2

In [72]:
class SimpleTokenizerV2:
    def __init__(self,vocab):
        self.str_to_int=vocab
        self.int_to_str = {integer:token for token,integer in vocab.items()}
        
    #Encoder returns the tokenIDs for the input text    
    def encode(self,text):
        preprocessed = re.split(r'([,.:;?_!"()\']|--|\s)',text)
        preprocessed = [item for item in preprocessed if item.strip()]
        preprocessed=[item if item in self.str_to_int else "<|unk|>" for item in preprocessed]
        ids = [self.str_to_int[token] for token in preprocessed]
        return ids
    
    #Decoder returns the output sentence where the input is the tokenIDs
    def decode(self,ids):
        output_string=" ".join([self.int_to_str[i] for i in ids])
        # Replace spaces before the specified punctuations
        output_string=re.sub(r'\s+(--|[,.?!"();\'])',r'\1',output_string)
        return output_string
        

##### The below tokenizer will work for the tokens out of the vocabulary.

In [76]:
text1="hello, world!"
text2="""I have mentioned-- that Mrs. Gisburn was rich; 
                 and it was immediately perceptible that her husband was 
                 extracting from this circumstance a delicate but substantial 
                 satisfaction. It is, as a rule, the people 
                 who scorn money who get most out of it; and 
                 Jack's elegant disdain of his wife's big balance enabled him, 
                 with an appearance of perfect good-breeding, to transmute it into objects of art and luxury."""
text=" <|endoftext|> ".join((text1,text2))
tokenizer=SimpleTokenizerV2(vocab)
ids=tokenizer.encode(text)
print(ids)

[1132, 5, 1132, 0, 1131, 53, 530, 667, 6, 987, 67, 7, 38, 1077, 841, 9, 157, 585, 1077, 567, 764, 987, 539, 562, 1077, 414, 477, 999, 266, 115, 323, 239, 947, 853, 7, 56, 584, 5, 177, 115, 848, 5, 988, 763, 1097, 859, 681, 1097, 488, 686, 738, 722, 585, 9, 157, 57, 2, 850, 384, 346, 722, 549, 1103, 2, 850, 219, 193, 387, 546, 5, 1108, 156, 166, 722, 765, 501, 5, 1016, 1027, 585, 579, 719, 722, 174, 157, 652, 7]


In [77]:
print(tokenizer.decode(ids))

<|unk|>, <|unk|>! <|endoftext|> I have mentioned-- that Mrs. Gisburn was rich; and it was immediately perceptible that her husband was extracting from this circumstance a delicate but substantial satisfaction. It is, as a rule, the people who scorn money who get most out of it; and Jack' s elegant disdain of his wife' s big balance enabled him, with an appearance of perfect good-breeding, to transmute it into objects of art and luxury.


<div class="alert alert-block alert-warning">

So far, we have discussed tokenization as an essential step in processing text as input to
LLMs. Depending on the LLM, some researchers also consider additional special tokens such
as the following:

[BOS] (beginning of sequence): This token marks the start of a text. It
signifies to the LLM where a piece of content begins.

[EOS] (end of sequence): This token is positioned at the end of a text,
and is especially useful when concatenating multiple unrelated texts,
similar to <|endoftext|>. For instance, when combining two different
Wikipedia articles or books, the [EOS] token indicates where one article
ends and the next one begins.

[PAD] (padding): When training LLMs with batch sizes larger than one,
the batch might contain texts of varying lengths. To ensure all texts have
the same length, the shorter texts are extended or "padded" using the
[PAD] token, up to the length of the longest text in the batch.

</div>


<div class="alert alert-info"> <b>Input:</b><br> Hello, how are you?<br><br>
<b>After adding [BOS]:</b><br> [BOS] Hello, how are you?

</div>

<div class="alert alert-info"> <b>Original:</b><br> Article 1: The sky is blue.<br> Article 2: Trees are green.<br><br>
<b>Tokenized with [EOS]:</b><br> The sky is blue. [EOS] Trees are green. [EOS]

</div>

<div class="alert alert-info"> <b>Sentences:</b><br> A: I like cats. (3 tokens)<br> B: Birds fly high. (3 tokens)<br> C: Wow! (1 token)<br><br>
<b>After padding:</b><br> A: I like cats.<br> B: Birds fly high.<br> C: Wow! [PAD] [PAD]

</div>

<div class="alert alert-block alert-warning">
GPT only uses <code>&lt;|endoftext|&gt;</code> token for simplicity
    </div>

<div class="alert alert-block alert-warning">
Even the tokenizer used for GPT models does not need <|unk|> tokens, instead they use byte pair encoding tokenizers, which each breaksdown words into subword units </div>

### Byte Pair Encoding

<div class="alert alert-block alert-success">
    <b>GPT2, GPT3 uses byte pair encoding </b>
</div>


In [78]:
!pip3 install tiktoken


[notice] A new release of pip is available: 23.3.2 -> 25.1
[notice] To update, run: C:\Users\joelb\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [81]:
import importlib
import tiktoken 

print("tiktoken version : ",importlib.metadata.version("tiktoken"))

tiktoken version :  0.3.3


In [82]:
#Intialising the BPE tokenizer from tiktoken
tokenizer = tiktoken.get_encoding("gpt2")

In [84]:
text="Hello, do you like tea? <|endoftext|> In the sunlit terraces of someUnknownPlace"
integers = tokenizer.encode(text,allowed_special={"<|endoftext|>"})
print(integers)

[15496, 11, 466, 345, 588, 8887, 30, 220, 50256, 554, 262, 4252, 18250, 8812, 2114, 286, 617, 20035, 27271]


In [85]:
strings=tokenizer.decode(integers)
print(strings)

Hello, do you like tea? <|endoftext|> In the sunlit terraces of someUnknownPlace


<div class="alert alert-block alert-warning">
We can make two noteworthy observations based on the token IDs and decoded text above.
First, the <code>&lt;|endoftext|&gt;</code> token is assigned a relatively large token ID, namely, 50256. </div>

<div class="alert alert-block alert-warning">
Second, the BPE tokenizer above encodes and decodes unknown words, such as "someunknownPlace" correctly.<br/> It can handle any unknown word. This was achieved not using <code>&lt;|unk|&gt;</code> keyword </div>

##### Another example for illustrating the usage of BPE tokenizer

In [87]:
integers = tokenizer.encode("Akir ewww")
print(integers)

strings=tokenizer.decode(integers)
print(strings)

[33901, 343, 304, 2503]
Akir ewww
